From 5b35757fc3613a18e2d784af53d1d23ebe295028 Mon Sep 17 00:00:00 2001
From: Rohan Garg <rohangarg@kubuntu.org>
Date: Fri, 31 Aug 2012 15:21:11 +0530
Subject: [PATCH] Make sure the plasma desktop scripts are sorted in the
 correct order

Instead of sorting the scripts according to their absolute paths,
which would cause scripts installed by packages to always end at
the top, scripts should be sorted by their relative path i.e.
plasma scripts from each directory that occurs in the path
should be sorted and then merged together to form a super list

REVIEWED BY: Marco Martin
---
 libs/plasmagenericshell/scripting/scriptengine.cpp |   20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/libs/plasmagenericshell/scripting/scriptengine.cpp b/libs/plasmagenericshell/scripting/scriptengine.cpp
index 777a56d..7810ab4 100644
--- a/libs/plasmagenericshell/scripting/scriptengine.cpp
+++ b/libs/plasmagenericshell/scripting/scriptengine.cpp
@@ -676,8 +676,24 @@ QStringList ScriptEngine::pendingUpdateScripts()
 QStringList ScriptEngine::defaultLayoutScripts()
 {
     const QString appName = KGlobal::activeComponent().aboutData()->appName();
-    QStringList scripts = KGlobal::dirs()->findAllResources("data", appName + "/init/*.js");
-    scripts.sort();
+    QStringList appNameDirs = KGlobal::dirs()->findDirs("data", appName);
+    QStringList scripts;
+    QDir appDir;
+    QFileInfoList scriptList;
+
+    foreach (const QString &appNameDir, appNameDirs) {
+        appDir.setPath(appNameDir + QLatin1String("init/"));
+        if (appDir.exists()) {
+            scriptList = appDir.entryInfoList(QStringList("*.js"),
+                                            QDir::NoFilter,
+                                            QDir::Name);
+            foreach (const QFileInfo &script, scriptList) {
+                if (script.exists()) {
+                    scripts.append(script.absoluteFilePath());
+                }
+            }
+        }
+    }
 
     QStringList scriptPaths;
 
-- 
1.7.10.4

